DO NOT MERGE MCOL-4516 Check for floating point rounding error in regr_**

davidhall586 opened the pull request:
Some floating point rounding error is inevitable during regr_** function execution. In some cases, when a intermediate result should be 0, it may become some microscopic negative number. This patch checks for this.

davidhall586 committed:
MCOL-4516 Check for floating point rounding error in regr_**
Some floating point rounding error is inevitable during regr_** function execution. In some cases, when a intermediate result should be 0, it may become some microscopic negative number. This patch checks for this.

davidhall586 committed:
MCOL-4516 Correct regr_sxx
In the first fix, the algorithm was wrong for sxx

abarkov reviewed (commented):
David, please add "HOW TO REPEAT" instructions into MCOL and corresponding mtr tests into this patch demonstrating that the problem does not exists any more.

abarkov commented on utils/regr/regr_intercept.cpp:
```diff
@@ -154,7 +154,7 @@ mcsv1_UDAF::ReturnCode regr_intercept::evaluate(mcsv1Context* context, static_an
         long double var_pop = (N * sumx2) - (sumx * sumx);
-        if (var_pop != 0)
+        if (var_pop != 0 && var_pop != -0)
         {
```
This change is meaningless I am afraid:
0 and -0 are integer (not DOUBLE!) literals and both are equal to INT(0), and then both are converted to DOUBLE for comparison.

So possibly you meant:

`if (var_pop != 0e0 && var_pop != -0e0)`

But this is not really needed because  0e0 and -0e0 are equal to each other although they are not binary equal.
This program proves:

```
#include <stdio.h>
#include <cstring>

int main()
{
  double a= -0e0;
  double b= +0e0;
  printf("memcmp=%d eq=%d\n", memcmp(&a, &b, sizeof(a)), a == b);
  return 0;
}
```
And the output is:
```
memcmp=128 eq=1
```

I guess the old version of this line was good enough.

davidhall586 replied:
Sometimes it appears to not compare properly. I don't know if this is an artifact of other bits being different or perhaps a float number so small in abs value that it prints as -0. Whatever the case, we can use the fact that var_pop should never be < 0.

davidhall586 replied:
After some research, it turns out var_pop cannot be negative, but covar_pop may be negative. The changes assume <0 should be 0 for var_pop and any negative values for covar_pop remain as calculated. This does allow for -0 to be printed in the answer for some functions.

abarkov commented on utils/regr/regr_slope.cpp:
```diff
@@ -154,7 +154,7 @@ mcsv1_UDAF::ReturnCode regr_slope::evaluate(mcsv1Context* context, static_any::a
         long double covar_pop = N * sumxy - sumx * sumy;
         long double var_pop = N * sumx2 - sumx * sumx;
-        if (var_pop != 0)
+        if (var_pop != 0 && var_pop != -0)
```
same here

abarkov commented on utils/regr/regrmysql.cpp:
```diff
@@ -537,7 +537,7 @@ extern "C"
             long double covar_pop = N * sumxy - sumx * sumy;
             long double var_pop = N * sumx2 - sumx * sumx;
-            if (var_pop != 0)
+            if (var_pop != 0 && var_pop != -0)
```
same here

abarkov commented on utils/regr/regrmysql.cpp:
```diff
@@ -657,7 +657,7 @@ extern "C"
             long double numerator = sumy * sumx2 - sumx * sumxy;
             long double var_pop = (N * sumx2) - (sumx * sumx);
-            if (var_pop != 0)
+            if (var_pop != 0 && var_pop != -0)
```
same here

abarkov commented on utils/regr/regrmysql.cpp:
```diff
@@ -1494,6 +1502,8 @@ extern "C"
             long double sumxy = data->sumxy;
             long double covar_samp = (sumxy - ((sumx * sumy) / N)) / (N - 1);
+            if (covar_samp < 0)  // might be -0
+                covar_samp = 0;
```
The above block is repeated in so many places. It would be nice to make a function or a class and reuse it everywhere.

davidhall586 replied:
Yeah. I thought about a sub-class and function, but decided I didn't want to spend the cpu cycles for the call. Could be done easily enough.

davidhall586 replied:
In this case covar_samp could have a legitimate negative value, so I removed the check

davidhall586 committed:
Merge branch 'develop' into MCOL-4516

davidhall586 committed:
Merge branch 'develop' into MCOL-4516

davidhall586 committed:
MCOL-4516 correct check for -0
var_pop is never negative. covar_pop may be negative
Adjusted the checks for -0 to this reality

davidhall586 committed:
Merge branch 'MCOL-4516' of https://github.com/dhall-MariaDB/mariadb-columnstore-engine into MCOL-4516

davidhall586 committed:
Merge branch 'develop' into MCOL-4516
